File: python/src/anonrouter_confidential/measurements.py
"""Load the operator-reviewed TEE/E2EE verification policy.

The pins live in the package-local ``measurements.json`` (a synced copy of
``shared/measurements.json``; DO NOT hand-edit the copy). A CI parity gate keeps
every copy byte-identical across the JS and Python SDKs. We LOAD this file at
import time via ``importlib.resources`` rather than hardcoding any pin.
"""

from __future__ import annotations

import json
from functools import lru_cache
from importlib import resources
from typing import Any

_PACKAGE = "anonrouter_confidential"
_FILENAME = "measurements.json"


@lru_cache(maxsize=1)
def load_measurements() -> dict[str, Any]:
    """Return the parsed measurements.json document (cached)."""
    text = resources.files(_PACKAGE).joinpath(_FILENAME).read_text(encoding="utf-8")
    data = json.loads(text)
    if not isinstance(data, dict):
        raise ValueError("measurements.json did not decode to an object")
    return data


def tdx_tee_type() -> int:
    """The pinned Intel TDX tee_type (129 == 0x81)."""
    value = load_measurements().get("tdxTeeType", 0x00000081)
    return int(value)


def pinned_measurement_policy_for(
    provider: str, upstream_model: str
) -> dict[str, Any] | None:
    """The verification policy for a route, or None when the provider binds
    the enclave by something other than a measurement policy (Venice).

    ``accepted`` is provider-specific: normally a measurement list, and for
    Tinfoil the fixed signed-release authority and repository.
    """
    providers = load_measurements().get("providers", {})
    entry = providers.get(provider)
    if not isinstance(entry, dict):
        return None

    if provider == "near-ai":
        by_model = entry.get("byModel", {})
        model_entry = by_model.get(upstream_model)
        if not isinstance(model_entry, dict):
            return None
        return _shape_policy(model_entry.get("measurementPolicy"))

    return _shape_policy(entry.get("measurementPolicy"))


def _shape_policy(policy: Any) -> dict[str, Any] | None:
    if not isinstance(policy, dict):
        return None
    return {
        "source": policy.get("source"),
        "version": policy.get("version"),
        "kind": policy.get("kind"),
        "accepted": policy.get("accepted", []),
    }


def pinned_endpoint_identity_for(
    provider: str, upstream_model: str
) -> str | None:
    """The operator-pinned direct endpoint identity for a route, if one exists."""
    providers = load_measurements().get("providers", {})
    entry = providers.get(provider)
    if not isinstance(entry, dict):
        return None
    if provider == "near-ai":
        model_entry = entry.get("byModel", {}).get(upstream_model)
        if isinstance(model_entry, dict):
            identity = model_entry.get("endpointIdentity")
            if isinstance(identity, str):
                return identity
        return None
    # Other providers (e.g. tinfoil) carry a single reviewed endpoint identity.
    identity = entry.get("endpointIdentity")
    return identity if isinstance(identity, str) else None

File: js/confidential/src/measurements.ts
// Loader for the operator-reviewed TEE/E2EE measurement policy. `measurements.json`
// is a byte-identical, CI-parity-gated copy of the monorepo's canonical
// `shared/measurements.json`, which holds the same policy AnonRouter's gateway
// enforces server-side. Static measurements remain code-reviewed pins. Tinfoil is
// different by design: its fixed policy pins the official signed-release authority
// and repository while the official verifier checks the current release output.
//
// Every verifier and E2EE transport reads its policy through this one loader, so
// there is a single source of truth for which enclave measurements and endpoints
// are acceptable.

import measurementsJson from "./measurements.json" with { type: "json" };

/** A single accepted TDX measurement identity (MRTD + RTMR0..3). NEAR entries
 *  additionally carry the pinned compose document digest + provenance. */
export interface TdxMeasurementEntry {
  name: string;
  mrTd: string;
  rtmr0: string;
  rtmr1: string;
  rtmr2: string;
  rtmr3: string;
  composeSha256?: string;
  composeRepository?: string;
  composeRepositoryCommit?: string;
  composeRepositoryPath?: string;
}

/** The one Tinfoil release authority accepted by the AnonRouter packages.
 *
 *  Only `configRepo` is re-derived from evidence: the Tinfoil verifier compares
 *  it against the repository the verification document names, so a different
 *  repository fails closed. `authority`, `releaseSelection` and
 *  `requireTaggedRelease` are LABELS recording which Tinfoil workflow this
 *  package was reviewed against, and they pin this file against a silent edit.
 *  Nothing here re-checks them: the document carries no Fulcio identity or
 *  workflow ref to check them against, and those guarantees belong to the
 *  official verifier's own `verifyCode` / `fetchDigest` steps.
 *
 *  Replaces the `TinfoilAcceptedRelease` entry type used through 0.1.1, which
 *  described a per-release fingerprint allowlist that no longer exists. */
export interface TinfoilProviderAuthority {
  authority: "github-actions-sigstore";
  configRepo: "tinfoilsh/confidential-model-router";
  releaseSelection: "latest";
  requireTaggedRelease: true;
}

/** The provider-neutral measurement policy the verifier binds evidence against.
 *  `accepted` is provider-specific and interpreted only by that provider verifier. */
export interface MeasurementPolicy {
  source: string;
  version: string;
  accepted: unknown;
}

interface RawMeasurementPolicy {
  source?: string;
  version?: string;
  kind?: string;
  accepted?: unknown;
}

interface RawMeasurements {
  schemaVersion?: number;
  tdxTeeType?: number;
  providers?: {
    chutes?: { endpointIdentity?: string; measurementPolicy?: RawMeasurementPolicy | null };
    tinfoil?: { endpointIdentity?: string; measurementPolicy?: RawMeasurementPolicy | null };
    venice?: { endpointIdentity?: string; measurementPolicy?: RawMeasurementPolicy | null };
    "near-ai"?: {
      byModel?: Record<string, { endpointIdentity?: string; measurementPolicy?: RawMeasurementPolicy | null }>;
    };
  };
}

const DATA = measurementsJson as unknown as RawMeasurements;

/** The Intel TDX `tee_type` value the pins were reviewed for (0x81 / 129). */
export const TDX_TEE_TYPE: number = typeof DATA.tdxTeeType === "number" ? DATA.tdxTeeType : 0x00000081;

function toPolicy(raw: RawMeasurementPolicy | null | undefined): MeasurementPolicy | undefined {
  if (!raw || typeof raw.source !== "string" || typeof raw.version !== "string") return undefined;
  return { source: raw.source, version: raw.version, accepted: raw.accepted ?? [] };
}

/**
 * Resolve the pinned measurement policy for a route, or undefined when none is
 * pinned (Venice binds on the signing key / workload keyset, not a measurement
 * allowlist, so it returns undefined by design). `near-ai` is keyed per upstream
 * model under `providers["near-ai"].byModel[model]`.
 */
export function pinnedMeasurementPolicyFor(provider: string, model: string): MeasurementPolicy | undefined {
  const providers = DATA.providers ?? {};
  if (provider === "chutes") return toPolicy(providers.chutes?.measurementPolicy);
  if (provider === "tinfoil") return toPolicy(providers.tinfoil?.measurementPolicy);
  if (provider === "venice") return toPolicy(providers.venice?.measurementPolicy);
  if (provider === "near-ai") return toPolicy(providers["near-ai"]?.byModel?.[model]?.measurementPolicy);
  return undefined;
}

/**
 * Resolve the operator-pinned endpoint identity (direct enclave host) for a route,
 * or undefined when the route is not host-pinned. Only `near-ai` direct routes are
 * host-pinned in policy today.
 */
export function pinnedEndpointIdentityFor(provider: string, model: string): string | undefined {
  const providers = DATA.providers ?? {};
  if (provider === "near-ai") {
    const identity = providers["near-ai"]?.byModel?.[model]?.endpointIdentity;
    return typeof identity === "string" ? identity : undefined;
  }
  // Other providers (e.g. tinfoil) carry a single reviewed endpoint identity.
  const entry = (providers as Record<string, { endpointIdentity?: string }>)[provider];
  return typeof entry?.endpointIdentity === "string" ? entry.endpointIdentity : undefined;
}

/** The raw, parsed policy document (for advanced callers that want the full pins). */
export function measurementPolicyDocument(): RawMeasurements {
  return DATA;
}
